hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/dn/ratis/TestDnRatisRandomRead.java

```diff
@@ -0,0 +1,203 @@
+        .setNumDatanodes(REPLICATION_COUNT)
+        .build();
+    cluster.waitForClusterToBeReady();
+
+    client = cluster.newClient();
+
+    OzoneBucket ozoneBucket =
+        TestDataUtil.createVolumeAndBucket(client, BUCKET_LAYOUT);
+    volumeName = ozoneBucket.getVolumeName();
+    bucketName = ozoneBucket.getName();
+
+    String rootPath = String.format("%s://%s.%s/",
+        OzoneConsts.OZONE_URI_SCHEME, bucketName, volumeName);
+
+    // Set the fs.defaultFS and start the filesystem
+    conf.set(CommonConfigurationKeysPublic.FS_DEFAULT_NAME_KEY, rootPath);
+    // Set the number of keys to be processed during batch operate.
+    conf.setInt(OZONE_FS_ITERATE_BATCH_SIZE, 5);
+
+    fs = FileSystem.get(conf);
+
+  }
+
+  @AfterClass
+  public static void teardown() {
+    IOUtils.closeQuietly(client);
+    if (cluster != null) {
+      cluster.shutdown();
+    }
+    IOUtils.closeQuietly(fs);
+  }
+
+  @Before
+  public void createVolumeAndBucket() throws IOException {
+    // create a volume and a bucket to be used by RootedOzoneFileSystem (OFS)
+    OzoneBucket bucket =
+        TestDataUtil.createVolumeAndBucket(client, BUCKET_LAYOUT);
+    volumeName = bucket.getVolumeName();
+    volumePath = new Path(OZONE_URI_DELIMITER, volumeName);
+    bucketName = bucket.getName();
+    bucketPath = new Path(volumePath, bucketName);
+  }
+
+  @After
+  public void cleanup() throws IOException {
+    fs.delete(bucketPath, true);
+    fs.delete(volumePath, false);
+  }
+
+  @Test
+  public void testRandomRead()
+      throws Exception {
+    String key = "ratiskeytest";
+    int numFiles = 100;
+    // write files
+    for (int i = 0; i < numFiles; i++) {
+      try (OzoneOutputStream outputStream = client.getObjectStore()
+          .getVolume(volumeName)
+          .getBucket(bucketName)
+          .createKey(key + i, i * 100,
+              RatisReplicationConfig.getInstance(
+                  HddsProtos.ReplicationFactor.THREE),
+              new HashMap<>())) {
+        outputStream.write(RandomUtils.nextBytes(i));
+      }
+    }
+    // read files
+    for (int i = 0; i < numFiles; i++) {
+      try (
+          OzoneInputStream ozoneInputStream = client.getObjectStore()
+              .getVolume(volumeName)
+              .getBucket(bucketName)
+              .readFile(key + i)) {
+        ozoneInputStream.read();
+      }
+    }
+    // collect metrics from datanode
+    List<Long> transactionsCountByDn = new ArrayList<>();
+    cluster.getHddsDatanodes().forEach(dn -> {
+      long metrics =
+          dn.getDatanodeStateMachine().getContainer().getMetrics()
+              .getContainerOpsMetrics(
+                  ContainerProtos.Type.GetBlock);
+      if (metrics > 0) {
+        transactionsCountByDn.add(metrics);
+      }
+      System.out.println(dn.getDatanodeDetails().getUuid() + " " + metrics);
+    });
+    // read transactions must exist in all three DN
+    assertEquals(REPLICATION_COUNT, transactionsCountByDn.size());
```

Comment:
For some reason this assertion is failing for me locally in all runs.

```
[ERROR] org.apache.hadoop.ozone.dn.ratis.TestDnRatisRandomRead.testRandomRead  Time elapsed: 22.193 s  <<< FAILURE!
java.lang.AssertionError: expected:<3> but was:<1>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:647)
	at org.junit.Assert.assertEquals(Assert.java:633)
	at org.apache.hadoop.ozone.dn.ratis.TestDnRatisRandomRead.testRandomRead(TestDnRatisRandomRead.java:194)
```

Reply:
Is it still actual?

Reply:
Yes.